keps/sig-storage/20190129-csi-migration.md

```diff
@@ -123,6 +127,57 @@ The detailed design was originally implemented as a [design proposal](https://gi
 
 * All volume operation paths covered by Migration Shim in Beta for >= 1 quarter without significant issues
 
+## Test Plan
+
+### Per-driver migration testing
+
+We will require each plugin/driver provider to set up public CI to run all existing
+in-tree plugin driver tests for their migrated driver.
+
+If migration is on for that plugin, we will add checks in the test framework
+that inspects CSI Drivers metrics to make sure that the driver is servicing the
+operations. We must also observe through metrics that none of the in-tree code
+is being called Additionally, we must test that a PV created from migrated
+dynamic provisioning is identical to the PV created from the in-tree plugin.
+
+The above is done by checking that no in-tree plugin code is emitting metrics
+when migration is on. We will also confirm that metrics are being emitted in
+general by confirming the existence of an indicator metric.
+
+This should cover all use cases of volume operations, including volume
+reconstruction.
+
+### Upgrade/Downgrade/Skew Testing
+
+We will have test clusters brought up that have different feature flags
+enabled on different components (ADC and Kubelet). Once these feature flag skew
+configurations are brought up the test itself would have to know what
+configuration it’s running in and validate the expected result.
+
+Configurations to test:
+
+| ADC               | Kubelet                                            | Expected Result                                                          |
+|-------------------|----------------------------------------------------|--------------------------------------------------------------------------|
+| ADC Migration On  | Kubelet Migration On                               | Fully migrated - result should be same as “Migration Shim Testing” above |
+| ADC Migration On  | Kubelet Migration Off (or Kubelet version too low) | No calls made to driver. All operations serviced by in-tree plugin       |
+| ADC Migration Off | Kubelet Migration On                               | Not supported config - Undefined behavior                                |
```

Comment:
is this possible in a downgrade scenario? or do nodes downgrade before control plane?

Reply:
nodes must be downgraded before the controller